src/Potato/Data/Text/Zipper2.hs

```diff
@@ -142,11 +141,31 @@ rightLeftWord = undefined
 
 -- | Move the cursor up one logical line (clearing the selection)
 up :: TextZipper -> TextZipper
-up z@(TextZipper lb b s a la) = undefined
```

Comment:
I think you can make several of these cases a lot easier if you deselect first and then recursively call up. Then all of the `(y:(ys:yss))` and `[y]` cases are handled by the `[]` cases.